File: itv/SergioDeIscar-itv-java/SergioDeIscar-itv-java/src/main/java/org/example/controllers/ItvController.java
package org.example.controllers;

import org.example.exceptions.ItvExceptions;
import org.example.interfaces.ItvExtension;
import org.example.models.Coche;
import org.example.models.Moto;
import org.example.models.Vehiculo;

import java.util.List;
import java.util.Map;

public class ItvController implements ItvExtension {
    private final ItvExtension repo;
    public ItvController(ItvExtension repo) {
        this.repo = repo;
    }
    @Override
    public Vehiculo save(Vehiculo entity) {
        checkVehiculo(entity);
        return repo.save(entity);
    }

    @Override
    public Vehiculo findById(String s) {
        checkMatricula(s);
        return repo.findById(s);
    }

    @Override
    public Vehiculo deleteById(String s) {
        checkMatricula(s);
        return repo.deleteById(s);
    }

    @Override
    public List<Vehiculo> findAll() {
        return repo.findAll();
    }

    @Override
    public void saveAll(List<Vehiculo> entities) {
        entities.forEach(v -> checkVehiculo(v));
        repo.saveAll(entities);
    }

    @Override
    public Boolean isFull() {
        return repo.isFull();
    }

    @Override
    public List<Vehiculo> getCoches() {
        return repo.getCoches();
    }

    @Override
    public Vehiculo getCocheMaxModerno() {
        return repo.getCocheMaxModerno();
    }

    @Override
    public Vehiculo getVehiculoMinKilometros() {
        return repo.getVehiculoMinKilometros();
    }

    @Override
    public List<Vehiculo> findByMarca(String marca) {
        return repo.findByMarca(marca);
    }

    @Override
    public Double getMediaKilometrosMotos() {
        return repo.getMediaKilometrosMotos();
    }

    @Override
    public Vehiculo getOldestCoche() {
        return repo.getOldestCoche();
    }

    @Override
    public Vehiculo getOldestCoche2Puertas() {
        return repo.getOldestCoche2Puertas();
    }

    @Override
    public Map<String, Integer> getCountVehiculos() {
        return repo.getCountVehiculos();
    }

    @Override
    public Map<String, Integer> getCountVehiculosAptos() {
        return repo.getCountVehiculosAptos();
    }

    @Override
    public Map<String, Double> getMediaVehiculosAnio() {
        return repo.getMediaVehiculosAnio();
    }

    @Override
    public Map<String, List<Vehiculo>> groupVehiculosByMarca() {
        return repo.groupVehiculosByMarca();
    }

    @Override
    public List<Vehiculo> sortVehiculosByAnio() {
        return repo.sortVehiculosByAnio();
    }

    @Override
    public List<Vehiculo> sortDesVehiculosByMarca() {
        return repo.sortDesVehiculosByMarca();
    }

    @Override
    public Map<String, List<Vehiculo>> groupVehiculosByMarcaSortByKilometro() {
        return repo.groupVehiculosByMarcaSortByKilometro();
    }

    private void checkVehiculo(Vehiculo vehiculo) {
        checkMatricula(vehiculo.getMatricula());
        if (checkNumber(vehiculo.getAnio().getYear(), 1900, 2023)) { throw new ItvExceptions.ItvAnioException(); }
        if(checkNumberMin(vehiculo.getKilometro(), 0)) { throw new ItvExceptions.ItvKilometroException(); }
        if (vehiculo instanceof Coche) {
            if (checkNumberMax(((Coche) vehiculo).getNumPuertas(), 7)) { throw new ItvExceptions.ItvNumPuertasException(); }
        }
        else if (vehiculo instanceof Moto) {
            if(checkNumberMin(((Moto) vehiculo).getCilindrada(), 100)) { throw new ItvExceptions.ItvCilindradaException(); }
        }
    }
    private void checkMatricula(String matricula) {
        if (!matricula.matches("[0-9]{4}[A-Z]{3}")) {
            throw new ItvExceptions.ItvMatriculaException();
        }
    }

    private Boolean checkNumberMax(int number, int max){
        return checkNumber(number, 0, max);
    }
    private Boolean checkNumberMin(int number, int min){
        return checkNumber(number, min, Integer.MAX_VALUE);
    }

    private Boolean checkNumber(int number, int min, int max){
        return number >= min && number <= max;
    }
}

File: itv/SergioDeIscar-itv-kotlin/SergioDeIscar-itv-kotlin/src/main/kotlin/controllers/ItvController.kt
package controllers

import exceptions.*
import interfaces.ItvExtension
import models.*

class ItvController(private val repository: ItvExtension): ItvExtension {
    override fun isFull(): Boolean {
        return repository.isFull()
    }

    override fun getCoches(): List<Vehiculo> {
        return repository.getCoches()
    }

    override fun getCocheMaxModerno(): Vehiculo {
        return repository.getCocheMaxModerno()
    }

    override fun getVehiculoMinKilometros(): Vehiculo {
        return repository.getVehiculoMinKilometros()
    }

    override fun findByMarca(marca: String): List<Vehiculo> {
        return repository.findByMarca(marca)
    }

    override fun getMediaKilometrosMotos(): Double {
        return repository.getMediaKilometrosMotos()
    }

    override fun getOldestCoche(): Vehiculo {
        return repository.getOldestCoche()
    }

    override fun getOldestCoche2Puertas(): Vehiculo {
        return repository.getOldestCoche2Puertas()
    }

    override fun getCountVehiculos(): Map<String, Int> {
        return repository.getCountVehiculos()
    }

    override fun getCountVehiculosAptos(): Map<String, Int> {
        return repository.getCountVehiculosAptos()
    }

    override fun getMediaVehiculosAnio(): Map<String, Double> {
        return repository.getMediaVehiculosAnio()
    }

    override fun groupVehiculosByMarca(): Map<String, List<Vehiculo>> {
        return repository.groupVehiculosByMarca()
    }

    override fun sortVehiculosByAnio(): List<Vehiculo> {
        return repository.sortVehiculosByAnio()
    }

    override fun sortDesVehiculosByMarca(): List<Vehiculo> {
        return repository.sortDesVehiculosByMarca()
    }

    override fun groupVehiculosByMarcaSortByKilometro(): Map<String, List<Vehiculo>> {
        return repository.groupVehiculosByMarcaSortByKilometro()
    }

    override fun save(t: Vehiculo): Vehiculo? {
        checkVehicle(t)
        return repository.save(t)
    }

    override fun find(id: String): Vehiculo? {
        checkMatricula(id)
        return repository.find(id)
    }

    override fun delete(id: String): Vehiculo? {
        checkMatricula(id)
        return repository.delete(id)
    }

    override fun getAll(): List<Vehiculo> {
        return repository.getAll()
    }

    override fun saveAll(list: List<Vehiculo>) {
        list.forEach { checkVehicle(it) }
        repository.saveAll(list)
    }

    private fun checkVehicle(vehiculo: Vehiculo) {
        checkMatricula(vehiculo.matricula)
        require(checkNum(vehiculo.kilometro)) { throw ItvKilometroException() }
        require(checkNum(vehiculo.anio.year, 1900, 2023)) { throw ItvAnioException() }

        when(vehiculo){
            is Coche -> require(checkNum(vehiculo.numPuertas, max = 7)) { throw ItvPlazasException() }
            is Moto -> require(checkNum(vehiculo.cilindrada, min = 100)) { throw ItvCilindradaException() }
        }
    }
    private fun checkMatricula(matricula: String){
        require(Regex("^[0-9]{4}[A-Z]{3}$").matches(matricula)) { throw ItvMatriculaException() }
    }
    private fun checkNum(num: Int, min: Int = 0, max: Int = Int.MAX_VALUE): Boolean{
        return num >= min && num <= max
    }
}